### PR TITLE
Feature/mindmeld converter

### PR DESCRIPTION
**Pull request by vijay120**
This PR changes the DF converter in the following ways:
1. Renames generated MM intents and entities without extraneous tokens like `entries` and `user_says`.
2. Adds DF converter test
3. Generates two commands, "app_specific" and "unrelated". All unspecified intents go into the "unrelated" domain.
4. Adds a new config.py file that contains parameters
5. Replaces the existing DF sample project with a Banking blueprint from DF.

### CODE REVIEW

File: tests/test_converter.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from mindmeld.test import TestConversation
```
Comment (minhtuev) [inline tests/test_converter.py]: move this file to https://github.com/cisco/mindmeld/blob/master/tests/converter/d

File: tests/conftest.py [hunk 2]
```diff
@@ -42,9 +43,16 @@
...
 )
-DIALOGFLOW_PROJECT_PATH = os.path.join(
+DF_PROJECT_PATH = os.path.join(
     os.path.dirname(os.path.abspath(__file__)), "converter/dialogflow_sample_project"
 )
+# This is the dialogflow app converted to mindmeld app
+MM_DF_PATH = os.path.join(
+    os.path.dirname(os.path.abspath(__file__)), "converter/mm_df_converted_project"
+)
+df_init = DialogflowConverter(DF_PROJECT_PATH, MM_DF_PATH)
```
Comment (minhtuev) [inline tests/conftest.py]: this should be inside a test

File: mindmeld/converter/dialogflow.py [hunk 3]
```diff
@@ -182,22 +186,24 @@ def _create_intents_directories(self, intents):
                 mindmeld_intent_directory_name = self.clean_check(
                     sub, self.intents_list
                 )
+
+                domain = 'unrelated' if \
+                    'default' in mindmeld_intent_directory_name else 'app_specific'
```
Comment (serapio) [inline mindmeld/converter/dialogflow.py]: if it has to split the line, we should just use a regular `if` block

File: mindmeld/converter/dialogflow.py [hunk 4]
```diff
@@ -225,12 +232,12 @@ def _create_intent_file(
...
                                 df_meta[1:],
                             )
+                            entity_type = self.clean_name(mm_meta) + "_" + language
 
-                        entity_type = self.clean_name(mm_meta) + "_entries_" + language
                         part = "{" + df_text + "|" + entity_type + "}"
                     else:
                         entity_type = (
-                            self.clean_name(df_meta[1:]) + "_entries_" + language
+                            self.clean_name(df_meta[1:]) + "_" + language
```
Comment (serapio) [inline mindmeld/converter/dialogflow.py]: are you guaranteed that `language` is a string?

File: mindmeld/converter/dialogflow.py [hunk 5]
```diff
@@ -379,26 +386,15 @@ def create_mindmeld_init(self):
...
-                                    ]
+                                function_name = "renameMe" + str(i) + "_" + language
+                                handles = [
+                                    "intent="
+                                    + "'"
+                                    + self.clean_name(datastore["name"])
+                                    + "_"
+                                    + language
+                                    + "'"
+                                ]
```
Comment (serapio) [inline mindmeld/converter/dialogflow.py]: f-string please

File: tests/conftest.py [hunk 6]
```diff
@@ -96,8 +104,17 @@ def mindmeld_project_path2():
...
+    return MM_DF_PATH
+
+
+@pytest.fixture(scope="session")
+def mm_df_nlp(mm_df_app_path):
+    """Provides a built processor instance"""
+    nlp = NaturalLanguageProcessor(app_path=mm_df_app_path)
+    nlp.build()
+    nlp.dump()
+    return nlp
```
Comment (serapio) [inline tests/conftest.py]: do these files get cleaned up after the test?

File: tests/converter/dialogflow_sample_project/intents/account.spending.check - context_ spending - date.json [hunk 7]
```diff
@@ -0,0 +1,71 @@
+{
```
Comment (serapio) [inline tests/converter/dialogflow_sample_project/intents/account.spending.check - context_ spending - date.json]: Are the spaces in the file name intentional?

File: mindmeld/converter/dialogflow.py (hunk 3 above)
Comment (vijay120) [reply mindmeld/converter/dialogflow.py]: done

File: mindmeld/converter/dialogflow.py (hunk 4 above)
Comment (vijay120) [reply mindmeld/converter/dialogflow.py]: Resolved this issue by making sure the function calling this function always has the language string set.

File: mindmeld/converter/dialogflow.py (hunk 5 above)
Comment (vijay120) [reply mindmeld/converter/dialogflow.py]: done

File: tests/test_converter.py (hunk 1 above)
Comment (vijay120) [reply tests/test_converter.py]: done

File: mindmeld/converter/generic_config.txt [hunk 8]
```diff
@@ -22,10 +22,52 @@ ENTITY_RECOGNIZER_CONFIG = {
...
+        },
+    },
+    "features": {
+        "bag-of-words": {"lengths": [1, 2, 3]},
+        "freq": {"bins": 5},
+        "enable-stemming": True,
+        "sys-candidates": {},
+    },
+}
+
```
Comment (arushir) [inline mindmeld/converter/generic_config.txt]: How was this config selected and why is it included? If it's meant to be the best default parameters for all new applications, we should update the default configs in the MindMeld platform instead

File: tests/conftest.py (hunk 2 above)
Comment (vijay120) [reply tests/conftest.py]: done

File: tests/conftest.py (hunk 6 above)
Comment (vijay120) [reply tests/conftest.py]: done

File: tests/converter/dialogflow_sample_project/intents/account.spending.check - context_ spending - date.json (hunk 7 above)
Comment (vijay120) [reply tests/converter/dialogflow_sample_project/intents/account.spending.check - context_ spending - date.json]: Yes, since this is the file name that is generated from DF

File: mindmeld/converter/generic_config.txt (hunk 8 above)
Comment (vijay120) [reply mindmeld/converter/generic_config.txt]: removed this file

File: mindmeld/converter/converter.py [hunk 9]
```diff
@@ -62,20 +62,6 @@ def create_directory(directory):
             except OSError:
                 logger.error("Cannot create directory at %s", directory)
 
-    @staticmethod
```
Comment (minhtuev) [inline mindmeld/converter/converter.py]: We should keep this function since it is still being used by the Rasa converter.

File: mindmeld/converter/dialogflow.py [hunk 10]
```diff
@@ -182,22 +186,27 @@ def _create_intents_directories(self, intents):
                 mindmeld_intent_directory_name = self.clean_check(
                     sub, self.intents_list
                 )
+
+                if 'default' in mindmeld_intent_directory_name:
```
Comment (minhtuev) [inline mindmeld/converter/dialogflow.py]: we should add documentation here

File: tests/converter/dialogflow_sample_project/stream.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/env python
```
Comment (minhtuev) [inline tests/converter/dialogflow_sample_project/stream.py]: Is this part of the DF flow sample project?

File: tests/converter/test_converter.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
+import os
```
Comment (minhtuev) [inline tests/converter/test_converter.py]: we should rename this file to `test_df_converter`

File: tests/converter/test_converter.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    mm_df_app = importlib.import_module('mm_df_converted_project').app
+    mm_df_app.lazy_init(mm_df_nlp)
+
+    conv = TestConversation(app=mm_df_app)
+    conv.process("what is my balance")
+    conv.assert_text("Here's your latest balance:")
+    conv.assert_domain("app_specific")
+    conv.assert_intent("accountbalancecheck_en")
+    conv.assert_frame({})
+
```
Comment (minhtuev) [inline tests/converter/test_converter.py]: we should also check in the nlp objects that:
- all domains, intents, entities are converted correctly
- the default domain is converted correctly

File: mindmeld/converter/test.py [hunk 14]
```diff
@@ -0,0 +1,7 @@
+from mindmeld.converter.dialogflow import DialogflowConverter
```
Comment (minhtuev) [inline mindmeld/converter/test.py]: we should consolidate this into `test_df_converter` to make sure that this function is always invoked before we can load and test the NLP app

File: mindmeld/converter/generic_config.txt [hunk 15]
```diff
@@ -1,31 +0,0 @@
-ENTITY_RECOGNIZER_CONFIG = {
```
Comment (minhtuev) [inline mindmeld/converter/generic_config.txt]: we should keep this since the rasa converter is also using this

File: mindmeld/converter/dialogflow.py (hunk 10 above)
Comment (vijay120) [reply mindmeld/converter/dialogflow.py]: done

File: tests/converter/dialogflow_sample_project/stream.py (hunk 11 above)
Comment (vijay120) [reply tests/converter/dialogflow_sample_project/stream.py]: good point, deleted.

File: tests/converter/test_converter.py (hunk 12 above)
Comment (vijay120) [reply tests/converter/test_converter.py]: done

File: mindmeld/converter/dialogflow.py (hunk 5 above)
Comment (minhtuev) [reply mindmeld/converter/dialogflow.py]: i don't think we can use fstring here since it is only available on python 3.6 and above while mindmeld still supports py3.5

File: tests/converter/test_converter.py (hunk 13 above)
Comment (minhtuev) [reply tests/converter/test_converter.py]: to clarify, for #3 we don't need to check all responses, but add a few more to make sure that the edge cases are addressed.

Comment (vijay120) [reply tests/converter/test_converter.py]: done with the first three points, point 4 is not supported in this PR

File: mindmeld/converter/generic_config.txt (hunk 15 above)
Comment (vijay120) [reply mindmeld/converter/generic_config.txt]: We should be using the default settings. The only difference between this and the default is stemming being enabled, which is not necessary for RASA or DF.

File: mindmeld/converter/test.py (hunk 14 above)
Comment (vijay120) [reply mindmeld/converter/test.py]: Deleted this file since the sample app has changed.

Comment (minhtuev) [reply mindmeld/converter/test.py]: should this file be deleted?

**Comment by arushir** [comment]
One more thought - since we are working on a MindMeld banking blueprint, maybe we could use the DF Hotel Booking or Car example applications instead? Just to give more diversity in examples.
